official/coin_build/man/man3/ScXMLDataElt.3
.TH "ScXMLDataElt" 3 "Sun May 28 2017" "Version 4.0.0a" "Coin" \" -*- nroff -*-
.ad l
.nh
.SH NAME
ScXMLDataElt \- the \fC<data>\fP SCXML element\&.  

.SH SYNOPSIS
.br
.PP
.PP
\fC#include <Inventor/scxml/ScXMLDataElt\&.h>\fP
.PP
Inherits \fBScXMLElt\fP\&.
.SS "Public Member Functions"

.in +1c
.ti -1c
.RI "virtual \fBSoType\fP \fBgetTypeId\fP (void) const"
.br
.RI "Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. "
.ti -1c
.RI "virtual void \fBsetIDAttribute\fP (const char *ID)"
.br
.ti -1c
.RI "virtual const char * \fBgetIDAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetSrcAttribute\fP (const char *src)"
.br
.ti -1c
.RI "virtual const char * \fBgetSrcAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual void \fBsetExprAttribute\fP (const char *expr)"
.br
.ti -1c
.RI "virtual const char * \fBgetExprAttribute\fP (void) const"
.br
.ti -1c
.RI "virtual SbBool \fBhandleXMLAttributes\fP (void)"
.br
.ti -1c
.RI "virtual void \fBcopyContents\fP (const \fBScXMLElt\fP *rhs)"
.br
.ti -1c
.RI "virtual const \fBScXMLElt\fP * \fBsearch\fP (const char *attrname, const char *attrvalue) const"
.br
.ti -1c
.RI "virtual void \fBsetValueObj\fP (\fBScXMLDataObj\fP *value)"
.br
.ti -1c
.RI "virtual \fBScXMLDataObj\fP * \fBgetValueObj\fP (void) const"
.br
.in -1c
.SS "Static Public Member Functions"

.in +1c
.ti -1c
.RI "static \fBScXMLEltReader\fP * \fBgetElementReader\fP (void)"
.br
.ti -1c
.RI "static \fBSoType\fP \fBgetClassTypeId\fP (void)"
.br
.ti -1c
.RI "static void * \fBcreateInstance\fP (void)"
.br
.ti -1c
.RI "static void \fBinitClass\fP (void)"
.br
.ti -1c
.RI "static void \fBcleanClass\fP (void)"
.br
.in -1c
.SS "Protected Attributes"

.in +1c
.ti -1c
.RI "char * \fBID\fP"
.br
.ti -1c
.RI "char * \fBsrc\fP"
.br
.ti -1c
.RI "char * \fBexpr\fP"
.br
.ti -1c
.RI "\fBScXMLDataObj\fP * \fBvalue\fP"
.br
.in -1c
.SS "Additional Inherited Members"
.SH "Detailed Description"
.PP 
the \fC<data>\fP SCXML element\&. 

The \fC<data>\fP SCXML element is for storing variables\&. They need to be contained within a \fC<datamodel>\fP element\&. \fC<data>\fP elements are only functional in state machines with a profile that has an implementation of the Data Module\&. The \fC'minimum'\fP profile does not, but the \fC'x-coin'\fP profile does\&.
.PP
The following attributes are accepted: 
.PD 0

.IP "\(bu" 2
\fCID\fP 
.IP "\(bu" 2
\fCsrc\fP - not supported 
.IP "\(bu" 2
\fCexpr\fP 
.PP
The \fCID\fP attribute is required and specified the variable name\&. The variable name must be unique for the whole SCXML state machine specification\&.
.PP
The \fCsrc\fP attribute is not supported\&.
.PP
The \fCexpr\fP attribute specifies the value of the variable\&. The attribute is evaluated based on which profile the state machine is using\&.
.PP
The value of the \fC<data>\fP variable can be changed through the use of \fC<assign>\fP elements, and their values can be passed on to other systems through the use of the \fI<send>\fP element in the External Communications Module\&.
.PP
\fBSince:\fP
.RS 4
Coin 4\&.0 
.RE
.PP
\fBSee also:\fP
.RS 4
\fBScXMLDataModelElt\fP, \fBScXMLAssignElt\fP, \fBScXMLSendElt\fP 
.RE
.PP

.SH "Member Function Documentation"
.PP 
.SS "\fBSoType\fP ScXMLDataElt::getTypeId (void) const\fC [virtual]\fP"

.PP
Returns the type identification of an object derived from a class inheriting \fBSoBase\fP\&. This is used for run-time type checking and 'downward' casting\&. 
.PP
\fBSee also:\fP
.RS 4
\fBSoBase::getTypeId() const\fP 
.RE
.PP

.PP
Implements \fBScXMLElt\fP\&.
.SS "SbBool ScXMLDataElt::handleXMLAttributes (void)\fC [virtual]\fP"
This method is called when the file reader has set all the XML attributes on the object, and wants the object to handle them and report if the values were ok or if a read error should be produced\&.
.PP
\fBReturns:\fP
.RS 4
TRUE if the attributes are ok, and FALSE on error\&.
.RE
.PP
This base class implementation does nothing and just returns TRUE\&. 
.PP
Reimplemented from \fBScXMLElt\fP\&.
.SS "const \fBScXMLElt\fP * ScXMLDataElt::search (const char * attrname, const char * attrvalue) const\fC [virtual]\fP"
This method searches the SCXML structure for an element with the given attributevalue for the given attribute\&.
.PP
Returns NULL if nothing was found\&.
.PP
This function needs to be reimplemented to traverse child elements\&. 
.PP
Reimplemented from \fBScXMLElt\fP\&.

.SH "Author"
.PP 
Generated automatically by Doxygen for Coin from the source code\&.


coin4/src/scxml/ScXMLDataElt.cpp
/**************************************************************************\
 * Copyright (c) Kongsberg Oil & Gas Technologies AS
 * All rights reserved.
 * 
 * Redistribution and use in source and binary forms, with or without
 * modification, are permitted provided that the following conditions are
 * met:
 * 
 * Redistributions of source code must retain the above copyright notice,
 * this list of conditions and the following disclaimer.
 * 
 * Redistributions in binary form must reproduce the above copyright
 * notice, this list of conditions and the following disclaimer in the
 * documentation and/or other materials provided with the distribution.
 * 
 * Neither the name of the copyright holder nor the names of its
 * contributors may be used to endorse or promote products derived from
 * this software without specific prior written permission.
 * 
 * THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS
 * "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT
 * LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR
 * A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT
 * HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL,
 * SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT
 * LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE,
 * DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY
 * THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT
 * (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE
 * OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
\**************************************************************************/

#include <Inventor/scxml/ScXMLDataElt.h>

/*!
  \class ScXMLDataElt ScXMLDataElt.h Inventor/scxml/ScXMLDataElt.h
  \brief the \c &lt;data&gt; SCXML element.

  The \c &lt;data&gt; SCXML element is for storing variables.  They need
  to be contained within a \c &lt;datamodel&gt; element.  \c &lt;data&gt;
  elements are only functional in state machines with a profile that has an
  implementation of the Data Module.  The \c "minimum" profile does not,
  but the \c "x-coin" profile does.

  The following attributes are accepted:
  \li \c ID
  \li \c src - not supported
  \li \c expr

  The \c ID attribute is required and specified the variable name.
  The variable name must be unique for the whole SCXML state machine
  specification.

  The \c src attribute is not supported.

  The \c expr attribute specifies the value of the variable.  The attribute
  is evaluated based on which profile the state machine is using.

  The value of the \c &lt;data&gt; variable can be changed through the use
  of \c &lt;assign&gt; elements, and their values can be passed on to other
  systems through the use of the \a &lt;send&gt; element in the External
  Communications Module.

  \since Coin 4.0
  \sa ScXMLDataModelElt, ScXMLAssignElt, ScXMLSendElt
  \ingroup scxml
*/

#include <cassert>
#include <cstring>
#include <vector>

#include <Inventor/errors/SoDebugError.h>
#include <Inventor/C/XML/element.h>

#include "scxml/ScXMLCommonP.h"
#include "SbBasicP.h"

// *************************************************************************

class ScXMLDataEltReader : public ScXMLEltReader {
public:
  ScXMLDataEltReader(void);
  virtual ScXMLElt * read(ScXMLElt * container, cc_xml_elt * xmlelt, ScXMLDocument * doc, ScXMLStateMachine * sm);
};

ScXMLDataEltReader::ScXMLDataEltReader(void)
: ScXMLEltReader("data")
{
}

ScXMLElt *
ScXMLDataEltReader::read(ScXMLElt * container, cc_xml_elt * xmlelt, ScXMLDocument * COIN_UNUSED_ARG(doc), ScXMLStateMachine * COIN_UNUSED_ARG(sm))
{
  assert(container && xmlelt);
  ScXMLDataElt * data = new ScXMLDataElt;
  data->setContainer(container);
  this->setXMLAttributes(data, xmlelt);

  // handle XML attributes
  if (unlikely(!data->handleXMLAttributes())) {
    delete data;
    return nullptr;
  }

  const int numchildren = cc_xml_elt_get_num_children(xmlelt);
  for (int c = 0; c < numchildren; ++c) {
    cc_xml_elt * element = cc_xml_elt_get_child(xmlelt, c);
    const char * elementtype = cc_xml_elt_get_type(element);

    if (std::strcmp(elementtype, COIN_XML_CDATA_TYPE) == 0) {
      // ignore CDATA
      continue;
    }

    SoDebugError::post("ScXMLDataEltReader::read",
                       "unexpected XML element '<%s>' found in <data>",
                       elementtype);
    delete data;
    return nullptr;
  }

  return data;
}

// *************************************************************************

class ScXMLDataElt::PImpl {
public:
};

#define PRIVATE(obj) ((obj)->pimpl)

SCXML_ELEMENT_SOURCE(ScXMLDataElt);

void
ScXMLDataElt::initClass(void)
{
  SCXML_OBJECT_INIT_CLASS(ScXMLDataElt, ScXMLElt, "ScXMLElt");
  SCXML_ELEMENT_REGISTER_READER(ScXMLDataElt, "data", ScXMLDataEltReader);
}

void
ScXMLDataElt::cleanClass(void)
{
  SCXML_ELEMENT_UNREGISTER_READER(ScXMLDataElt);
  ScXMLDataElt::classTypeId = SoType::badType();
}

ScXMLDataElt::ScXMLDataElt(void)
: ID(nullptr),
  src(nullptr),
  expr(nullptr)
{
}

ScXMLDataElt::~ScXMLDataElt(void)
{
  this->setIDAttribute(nullptr);
  this->setSrcAttribute(nullptr);
  this->setExprAttribute(nullptr);
}

void
ScXMLDataElt::setIDAttribute(const char * IDstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->ID, "ID", IDstr);
}

// const char * getIDAttribute(void) const { return this->ID; }

void
ScXMLDataElt::setSrcAttribute(const char * srcstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->src, "src", srcstr);
}

// const char * getSrcAttribute(void) const { return this->src; }

void
ScXMLDataElt::setExprAttribute(const char * exprstr)
{
  SCXML__SET_ATTRIBUTE_VALUE(this->expr, "expr", exprstr);
}

// const char * getExprAttribute(void) const { return this->expr; }

bool
ScXMLDataElt::handleXMLAttributes(void)
{
  if (!inherited::handleXMLAttributes()) {
    return false;
  }

  this->setIDAttribute(this->getXMLAttribute("ID"));
  this->setSrcAttribute(this->getXMLAttribute("src"));
  this->setExprAttribute(this->getXMLAttribute("expr"));

  return true;
}

void
ScXMLDataElt::copyContents(const ScXMLElt * rhs)
{
  inherited::copyContents(rhs);
  const ScXMLDataElt * orig = coin_assert_cast<const ScXMLDataElt *>(rhs);
  this->setIDAttribute(orig->getIDAttribute());
  this->setSrcAttribute(orig->getSrcAttribute());
  this->setExprAttribute(orig->getExprAttribute());
}

const ScXMLElt *
ScXMLDataElt::search(const char * attrname, const char * attrvalue) const
{
  const ScXMLElt * hit = inherited::search(attrname, attrvalue);
  if (hit) {
    return hit;
  }
  if (std::strcmp(attrname, "ID") == 0) {
    if (this->ID && std::strcmp(attrvalue, this->ID) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "src") == 0) {
    if (this->src && std::strcmp(attrvalue, this->src) == 0) {
      return this;
    }
  }
  else if (std::strcmp(attrname, "expr") == 0) {
    if (this->expr && std::strcmp(attrvalue, this->expr) == 0) {
      return this;
    }
  }
  return nullptr;
}

void
ScXMLDataElt::setValueObj(ScXMLDataObj * obj)
{
  this->value = obj;
}

#undef PRIVATE
